File: Leetcode/229-Majority-Element-II/MajorityElementII.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var majorityElement = function(nums) {
    var res = []
    if (!nums || nums.length == 0) {
        return res
    }
    if (nums.length < 3) {
        for (index in nums) {
            if (!res.includes(nums[index])) {
                res.push(nums[index])
            }
        }
        return res
    }
    var number1
    var number2
    var count1 = 0
    var count2 = 0
    for (index in nums) {
        if (number1 == nums[index]) {
            count1++
        } else if (number2 == nums[index]) {
            count2++
        } else if (count1 == 0) {
            number1 = nums[index]
            count1 = 1
        } else if (count2 == 0) {
            number2 = nums[index]
            count2 = 1
        } else {
            count1--
            count2--
        }
    }
    
    count1 = count2 = 0
    for (index in nums) {
        if (number1 == nums[index]) {
            count1++
        }
        if (number2 == nums[index]) {
            count2++
        }
    }
    if (count1 > nums.length/3) {
        res.push(number1)
    }
    if (count2 > nums.length/3) {
        res.push(number2)
    }
    return res
};

File: Leetcode/229-Majority-Element-II/MajorityElementII.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> res;
        int number1=0;
        int number2=0;
        int count1=0;
        int count2=0;
        int n = nums.size();
        for (int num:nums) {
            if (num==number1) {
                count1++;
            } else if (num==number2) {
                count2++;
            } else if (count1==0) {
                number1 = num;
                count1 = 1;
            } else if (count2==0) {
                number2 = num;
                count2 = 1;
            } else {
                count1--;
                count2--;
            }
        }
        count1 = count2 = 0;
        for (int num:nums) {
            if (num == number1) {
                count1++;
            } else if (num == number2) {
                count2++;
            }
        }
        if (count1 > n/3) {
            res.push_back(number1);
        }
        if (count2 > n/3) {
            res.push_back(number2);
        }
        return res;
    }
};


